modules/custom/falcon_mail/src/Plugin/Mail/FalconMailSystem.php

```diff
@@ -0,0 +1,138 @@
+ *
+ * @Mail(
+ *   id = "falcon_mail_system",
+ *   label = @Translation("Falcon Mailer"),
+ *   description = @Translation("Provides formatter that can wrap emails into HTML templates and replace tokens.")
+ * )
+ */
+class FalconMailSystem extends PhpMail {
+
+  /**
+   * Concatenate and wrap the e-mail body for either HTML emails.
+   *
+   * @param array $message
+   *   A message array.
+   *
+   * @return array
+   *   The formatted $message.
+   */
+  public function format(array $message) {
+    $message['subject'] = !empty($message['subject']) ? $message['subject'] :
+      (isset($message['params']['subject']) ? $message['params']['subject'] : '');
+    $message['body'] = !empty($message['body']) ? $message['body'] :
+      (isset($message['params']['body']) ? $message['params']['body'] : []);
+    $message['from'] = !empty($message['from']) ? $message['from'] :
+      (isset($message['params']['from']) ? $message['params']['from'] : Drupal::config('system.site')->get('mail'));
+
+    // Merge headers.
+    if (!empty($message['params']['headers'])) {
+      $message['headers'] = array_merge($message['headers'], $message['params']['headers']);
+    }
+
+    if (is_array($message['body'])) {
+      $message['body'] = implode("\n\n", $message['body']);
+    }
+
+    // Wrap body in theme template.
+    if (!empty($message['params']['theme_template'])) {
+      $message['body'] = str_replace('#$#BODY#$#', $message['body'], $message['params']['theme_template']);
+    }
+
+    // Replace tokens.
+    if (!empty($message['params']['replace_tokens'])) {
+
+      $tokens = !empty($message['params']['render_tokens']) ? $message['params']['render_tokens'] : [];
+      $token_options = !empty($message['params']['token_options']) ? $message['params']['token_options'] : [];
+
+      // Replace tokens and set new body.
+      $message['body'] = Drupal::token()->replace($message['body'], $tokens, $token_options);
+    }
+
+    // Replace relative urls in body.
+    $message['body'] = $this->replaseRelativeUrlsWithSaveStyles($message['body']);
+
+    return $message;
+  }
+
+  /**
+   * Send the e-mail message with using default PHP_Mail.
+   *
+   * @param array $message
+   *   A message array.
+   *
+   * @return bool
+   *   TRUE if the mail was successfully accepted, otherwise FALSE.
+   */
+  public function mail(array $message) {
+    return parent::mail($message);
+  }
+
+  /**
+   * Replace relative urls on absolute with saving <html>, <body>, <head> tags.
+   *
+   * @param string $body
+   *   Mail body.
+   *
+   * @return string
+   *   String with/without replaced relative urls.
+   */
+  public function replaseRelativeUrlsWithSaveStyles($body) {
+    try {
+      // Create Global dom document.
+      $html = new \DOMDocument();
+      $html->loadHTML($body);
+
+      // Get body element.
+      $bodyElement = $html->getElementsByTagName('body')[0];
+
+      if (empty($bodyElement)) {
```

Comment:
I'd first check that `$bodyElement->count())` is greater than zero and only after that assigned `$html->getElementsByTagName('body')[0]` to a variable.

Reply:
Changed.